File: backend-express/src/utils/symbolMapper.ts
/**
 * Symbol mapper for trading platform.
 * Maps trading symbols to script tokens for Angel One API compatibility.
 * TypeScript version of Python symbol_mapper.py
 */

// Exchange Types
export const EXCHANGE_NSE = 1;
export const EXCHANGE_NFO = 2;
export const EXCHANGE_BSE = 3;

export interface SymbolInfo {
    token: string;
    exchange: number;
    tradingSymbol: string;
    name: string;
}

// Symbol Token Mapping for NSE Equity
export const SYMBOL_TOKEN_MAP: Record<string, SymbolInfo> = {
    'RELIANCE': {
        token: '2885',
        exchange: EXCHANGE_NSE,
        tradingSymbol: 'RELIANCE-EQ',
        name: 'Reliance Industries Ltd'
    },
    'TCS': {
        token: '11536',
        exchange: EXCHANGE_NSE,
        tradingSymbol: 'TCS-EQ',
        name: 'Tata Consultancy Services Ltd'
    },
    'INFY': {
        token: '1594',
        exchange: EXCHANGE_NSE,
        tradingSymbol: 'INFY-EQ',
        name: 'Infosys Ltd'
    },
    'HDFCBANK': {
        token: '1333',
        exchange: EXCHANGE_NSE,
        tradingSymbol: 'HDFCBANK-EQ',
        name: 'HDFC Bank Ltd'
    },
    'ICICIBANK': {
        token: '4963',
        exchange: EXCHANGE_NSE,
        tradingSymbol: 'ICICIBANK-EQ',
        name: 'ICICI Bank Ltd'
    },
    'ITC': {
        token: '1660',
        exchange: EXCHANGE_NSE,
        tradingSymbol: 'ITC-EQ',
        name: 'ITC Ltd'
    },
    'SBIN': {
        token: '3045',
        exchange: EXCHANGE_NSE,
        tradingSymbol: 'SBIN-EQ',
        name: 'State Bank of India'
    },
    'BHARTIARTL': {
        token: '10604',
        exchange: EXCHANGE_NSE,
        tradingSymbol: 'BHARTIARTL-EQ',
        name: 'Bharti Airtel Ltd'
    },
    'HINDUNILVR': {
        token: '1394',
        exchange: EXCHANGE_NSE,
        tradingSymbol: 'HINDUNILVR-EQ',
        name: 'Hindustan Unilever Ltd'
    },
    'LT': {
        token: '11483',
        exchange: EXCHANGE_NSE,
        tradingSymbol: 'LT-EQ',
        name: 'Larsen & Toubro Ltd'
    },
};

// Reverse mapping: Token ID -> Symbol
export const TOKEN_TO_SYMBOL: Record<string, string> = Object.entries(SYMBOL_TOKEN_MAP)
    .reduce((acc, [symbol, info]) => {
        acc[info.token] = symbol;
        return acc;
    }, {} as Record<string, string>);

/**
 * Get script token from symbol name
 */
export function getTokenFromSymbol(symbol: string): string | null {
    const info = SYMBOL_TOKEN_MAP[symbol.toUpperCase()];
    return info ? info.token : null;
}

/**
 * Get symbol name from token ID
 */
export function getSymbolFromToken(token: string): string | null {
    return TOKEN_TO_SYMBOL[token] || null;
}

/**
 * Get full symbol info
 */
export function getSymbolInfo(symbol: string): SymbolInfo | null {
    return SYMBOL_TOKEN_MAP[symbol.toUpperCase()] || null;
}

/**
 * Get all available symbols
 */
export function getAllSymbols(): string[] {
    return Object.keys(SYMBOL_TOKEN_MAP);
}

/**
 * Check if a symbol is valid/supported
 */
export function isValidSymbol(symbol: string): boolean {
    return symbol.toUpperCase() in SYMBOL_TOKEN_MAP;
}

File: data-engine/symbol_mapper.py
"""
Symbol mapper for Angel One SmartAPI.
Maps trading symbols to Angel One token IDs for WebSocket subscription.
"""

# Exchange Types
EXCHANGE_NSE = 1
EXCHANGE_NFO = 2
EXCHANGE_BSE = 3

# Symbol Token Mapping for NSE Equity
# Token IDs are obtained from Angel One's instrument master file
# These are the most commonly traded stocks on NSE
SYMBOL_TOKEN_MAP = {
    'RELIANCE': {
        'token': '2885',
        'exchange': EXCHANGE_NSE,
        'trading_symbol': 'RELIANCE-EQ',
        'name': 'Reliance Industries Ltd'
    },
    'TCS': {
        'token': '11536',
        'exchange': EXCHANGE_NSE,
        'trading_symbol': 'TCS-EQ',
        'name': 'Tata Consultancy Services Ltd'
    },
    'INFY': {
        'token': '1594',
        'exchange': EXCHANGE_NSE,
        'trading_symbol': 'INFY-EQ',
        'name': 'Infosys Ltd'
    },
    'HDFCBANK': {
        'token': '1333',
        'exchange': EXCHANGE_NSE,
        'trading_symbol': 'HDFCBANK-EQ',
        'name': 'HDFC Bank Ltd'
    },
    'ICICIBANK': {
        'token': '4963',
        'exchange': EXCHANGE_NSE,
        'trading_symbol': 'ICICIBANK-EQ',
        'name': 'ICICI Bank Ltd'
    },
    'ITC': {
        'token': '1660',
        'exchange': EXCHANGE_NSE,
        'trading_symbol': 'ITC-EQ',
        'name': 'ITC Ltd'
    },
    'SBIN': {
        'token': '3045',
        'exchange': EXCHANGE_NSE,
        'trading_symbol': 'SBIN-EQ',
        'name': 'State Bank of India'
    },
    'BHARTIARTL': {
        'token': '10604',
        'exchange': EXCHANGE_NSE,
        'trading_symbol': 'BHARTIARTL-EQ',
        'name': 'Bharti Airtel Ltd'
    },
    'HINDUNILVR': {
        'token': '1394',
        'exchange': EXCHANGE_NSE,
        'trading_symbol': 'HINDUNILVR-EQ',
        'name': 'Hindustan Unilever Ltd'
    },
    'LT': {
        'token': '11483',
        'exchange': EXCHANGE_NSE,
        'trading_symbol': 'LT-EQ',
        'name': 'Larsen & Toubro Ltd'
    },
}

# Reverse mapping: Token ID -> Symbol
TOKEN_TO_SYMBOL = {info['token']: symbol for symbol, info in SYMBOL_TOKEN_MAP.items()}


def get_token_list_for_subscription() -> list[dict]:
    """
    Get token list formatted for WebSocket V2 subscription.
    Returns list of dicts with exchangeType and tokens.
    """
    # Group tokens by exchange
    exchange_tokens = {}
    for symbol, info in SYMBOL_TOKEN_MAP.items():
        exchange = info['exchange']
        if exchange not in exchange_tokens:
            exchange_tokens[exchange] = []
        exchange_tokens[exchange].append(info['token'])
    
    # Format for WebSocket subscription
    return [
        {'exchangeType': exchange, 'tokens': tokens}
        for exchange, tokens in exchange_tokens.items()
    ]


def get_symbol_from_token(token: str) -> str | None:
    """Get symbol name from token ID."""
    return TOKEN_TO_SYMBOL.get(token)


def get_token_info(symbol: str) -> dict | None:
    """Get full token info for a symbol."""
    return SYMBOL_TOKEN_MAP.get(symbol.upper())


def get_all_symbols() -> list[str]:
    """Get list of all tracked symbols."""
    return list(SYMBOL_TOKEN_MAP.keys())


def get_token_from_symbol(symbol: str) -> str | None:
    """Get token ID from symbol name."""
    info = SYMBOL_TOKEN_MAP.get(symbol.upper())
    return info['token'] if info else None

